docs/source/concepts/local-ledger.rst

```diff
@@ -0,0 +1,452 @@
+
+**NoSuchKey** actions are similar to non-consuming **Exercise**\ s and **Fetch**\ es of contracts when it comes to causal ordering: If there were another transaction `tx5'` with a **NoSuchKey** `(Acc, Bank, P)` action, then `tx5` and `tx5'` need not be ordered, just like `tx2` and `tx3` are unordered.
+
+.. _causality-consistency-ledger-model:
+
+From causality graphs to ledgers
+================================
+
+Since causality graphs are acyclic, their vertices can be sorted topologically and the resulting list is again a causality graph, where every vertex has an outgoing edge to all later vertices.
+If the original causality graph is `X`\ -consistent, then so is the topological sort, as topological sorting merely adds edges.
+For example, the transactions on the :ref:`ledger <split-counteroffer-ledger>` in the :ref:`out-of-band causality example <causality-example-out-of-band>` are a topological sort of the :ref:`corresponding causality graph <causality-graph-couteroffer-split>`.
+
+Conversely, we can reduce an `X`\ -consistent causality graph to only the causal dependencies that `X`\ -consistency imposes.
+This gives a minimal `X`\ -consistent causality graph.
+
+.. _def-reduction-causality-graph:
+
+Definition »Reduction of a consistent causality graph«
+  For an `X`\ -consistent causality graph `G`, there exists a unique minimal `X`\ -consistent causality graph `reduce`:sub:`X`\ `(G)` with the same vertices and the edges being a subset of `G`.
+  `reduce`:sub:`X`\ `(G)` is called the `X`\ -**reduction** of `G`.
+  As before, `X` is omitted if it contains all actions in `G`.
+
+The causality graph for the `CounterOffer` workflow is minimal and therefore its own reduction.
+It is also the reduction of the topological sort, i.e., the :ref:`ledger <split-counteroffer-ledger>` in the :ref:`out-of-band causality example <causality-example-out-of-band>`.
+
+.. note::
+   The reduction `reduce`:sub:`X`\ `(G)` of an `X`\ -consistent causality graph `G` can be computed as follows:
+   
+   #. Set the vertices of `G'` to the vertices of `G`.
+   #. The causal consistency conditions for contracts and keys demand that certain pairs of actions
+      `act`:sub:`1` and `act`:sub:`2` in `X` must be action-ordered.
+      For each such pair, determine the actions' ordering in `G` and add an edge to `G'` from the earlier action's transaction to the later action's transaction.
+   #. `reduce`:sub:`X`\ `(G)` is the transitive closure of `G'`.
+
+Topological sort and reduction link causality graphs `G` to the ledgers `L` from the DAML Ledger Model.
+Topological sort transforms a causality graph `G` into a sequence of transactions `topo(G)`; extending them with the requesters gives a sequence of commits, i.e., a ledger in the DAML Ledger Model.
```

Comment:
Minor: notation `topo(-)` not defined before. Is it a function or a relation?

Reply:
I read this as being the definition of `topo` as in " For a causality graph `G`, `topo(G)` is the topological sort of `G`"

Reply:
It was meant as @bame-da writes. But we actually use this notation only once more, so I removed it completely.